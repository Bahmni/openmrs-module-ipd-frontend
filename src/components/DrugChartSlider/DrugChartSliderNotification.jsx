import React from "react";
import PropTypes from "prop-types";
import { NotificationCarbon } from "bahmni-carbon-ui";
import { FormattedMessage } from "react-intl";
import { I18nProvider } from "../../features/i18n/I18nProvider";
import "./DrugChartSliderNotification.scss";

export default function DrugChartSliderNotification(props) {
  const { hostData, hostApi } = props;
  const getMessage = () => {
    switch (hostData?.notificationKind) {
      case "success":
        return (
          <I18nProvider>
            <FormattedMessage id="DRUG_CHART_MODAL_SAVE_MESSAGE" />
          </I18nProvider>
        );
      case "warning":
        return (
          <I18nProvider>
            <FormattedMessage id="DRUG_CHART_MODAL_CANCEL_MESSAGE" />
          </I18nProvider>
        );
    }
  };
  const title = getMessage();

  return (
    <NotificationCarbon
      className="notification"
      showMessage={true}
      title={title}
      kind={hostData?.notificationKind}
      onClose={() => {
        hostApi?.onClose();
      }}
      hideCloseButton={true}
    />
  );
}

DrugChartSliderNotification.propTypes = {
  hostData: PropTypes.shape({
    notificationKind: PropTypes.string.isRequired,
  }),
  hostApi: PropTypes.shape({
    onClose: PropTypes.func,
  }).isRequired,
};
