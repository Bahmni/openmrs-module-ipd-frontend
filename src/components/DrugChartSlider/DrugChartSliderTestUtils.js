export const mockScheduleDrugOrder = {
  uuid: "e2b966e0-1d5f-11e0-b929-000c29ad1d07",
  dateActivated: "2010-12-22T00:00:00.000+0530",
  drug: {
    uuid: "e2b83a8c-1d5f-11e0-b929-000c29ad1d07",
    name: "Aspirin",
  },
  duration: 10,
  durationUnit: "Days",
  uniformDosingType: {
    dose: 10,
    doseUnits: "Tablet",
    frequency: "Twice a day",
  },
  instructions: "",
  additionalInstructions: "",
  route: "Oral",
};

export const mockStartTimeDrugOrder = {
  uuid: "e2b966e0-1d5f-11e0-b929-000c29ad1d07",
  dateActivated: "2010-12-22T00:00:00.000+0530",
  drug: {
    uuid: "e2b83a8c-1d5f-11e0-b929-000c29ad1d07",
    name: "Aspirin",
  },
  duration: 10,
  durationUnit: "Days",
  uniformDosingType: {
    dose: 10,
    doseUnits: "Tablet",
    frequency: "Once a day",
  },
  instructions: "",
  additionalInstructions: "",
  route: "Oral",
};

export const mockScheduleFrequencies = [
  {
    name: "Twice a day",
    frequencyPerDay: 2,
  },
  {
    name: "Thrice a day",
    frequencyPerDay: 3,
  },
  {
    name: "Four times a day",
    frequencyPerDay: 4,
  },
];

export const mockScheduleFrequenciesWithTimings = [
  {
    name: "Twice a day",
    frequencyPerDay: 2,
    scheduleTiming: ["8:00", "16:00"],
  },
  {
    name: "Thrice a day",
    frequencyPerDay: 3,
    scheduleTiming: ["8:00", "10:00", "20:00"],
  },
  {
    name: "Four times a day",
    frequencyPerDay: 4,
    scheduleTiming: ["8:00", "10:00", "14:00", "20:00"],
  },
];

export const mockStartTimeFrequencies = [
  "Every Hour",
  "Every 2 hours",
  "Every 3 hours",
  "Every 4 hours",
  "Every 6 hours",
  "Every 8 hours",
  "Every 12 hours",
  "Once a day",
  "Nocte (At Night)",
  "Every 30 minutes",
];

export const mockDrugOrderFrequencies = [
  {
    uuid: "9d7c32a2-3f10-11e4-adec-0800271c1b75",
    frequencyPerDay: 1,
    name: "Once a day",
  },
  {
    uuid: "9d7d0641-3f10-11e4-adec-0800271c1b75",
    frequencyPerDay: 2,
    name: "Twice a day",
  },
  {
    uuid: "9d838148-3f10-11e4-adec-0800271c1b75",
    frequencyPerDay: 3,
    name: "Thrice a day",
  },
  {
    uuid: "9d84890a-3f10-11e4-adec-0800271c1b75",
    frequencyPerDay: 4,
    name: "Four times a day",
  },
  {
    uuid: "33443dad-8a92-11e4-977f-0800271c1b75",
    frequencyPerDay: 24,
    name: "Every Hour",
  },
  {
    uuid: "3345a0d3-8a92-11e4-977f-0800271c1b75",
    frequencyPerDay: 12,
    name: "Every 2 hours",
  },
  {
    uuid: "334c9330-8a92-11e4-977f-0800271c1b75",
    frequencyPerDay: 8,
    name: "Every 3 hours",
  },
  {
    uuid: "334d6ca5-8a92-11e4-977f-0800271c1b75",
    frequencyPerDay: 6,
    name: "Every 4 hours",
  },
  {
    uuid: "334e5025-8a92-11e4-977f-0800271c1b75",
    frequencyPerDay: 4,
    name: "Every 6 hours",
  },
  {
    uuid: "334f3890-8a92-11e4-977f-0800271c1b75",
    frequencyPerDay: 3,
    name: "Every 8 hours",
  },
  {
    uuid: "33500c00-8a92-11e4-977f-0800271c1b75",
    frequencyPerDay: 2,
    name: "Every 12 hours",
  },
  {
    uuid: "0",
    frequencyPerDay: 1,
    name: "Immediately",
  },
  {
    uuid: "8545fc4a-310c-11ee-bcc6-0242c0a82008",
    frequencyPerDay: 1,
    name: "Nocte (At Night)",
  },
  {
    uuid: "85494fb0-310c-11ee-bcc6-0242c0a82008",
    frequencyPerDay: 48,
    name: "Every 30 minutes",
  },
];
