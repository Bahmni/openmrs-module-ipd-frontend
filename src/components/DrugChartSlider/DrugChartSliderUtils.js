import moment from "moment";

export const isInvalidTimeTextPresent = (
  enable24HourTimers,
  invalidTimeText12Hour,
  invalidTimeText24Hour
) => {
  const screenContent = document.body.textContent;
  const invalidTimeText = enable24HourTimers
    ? invalidTimeText24Hour
    : invalidTimeText12Hour;
  return screenContent.includes(invalidTimeText);
};

export const isTimePassed = (newTime, enable24HourTimers) => {
  const currentTime = moment();
  const enteredTime = enable24HourTimers
    ? moment(newTime, "HH:mm")
    : moment(newTime, "hh:mm A");
  return currentTime.isAfter(enteredTime);
};

export const updateStartTimeBasedOnFrequency = (frequency, time) => {
  switch (frequency) {
    case "Every Hour":
      time.add(1, "hour");
      break;
    case "Every 2 hours":
      time.add(2, "hours");
      break;
    case "Every 3 hours":
      time.add(3, "hours");
      break;
    case "Every 4 hours":
      time.add(4, "hours");
      break;
    case "Every 6 hours":
      time.add(6, "hours");
      break;
    case "Every 8 hours":
      time.add(8, "hours");
      break;
    case "Every 12 hours":
      time.add(12, "hours");
      break;
    case "Once a day":
      time.add(1, "day");
      break;
    case "Nocte (At Night)":
      time.set({ hour: 23, minute: 59, second: 59 });
      break;
    case "Every 30 minutes":
      time.add(30, "minutes");
      break;
    default:
      break;
  }
  return time;
};

const areSchedulesInOrder = (allSchedule) => {
  return allSchedule.every((schedule, index) => {
    if (index === 0) return true;
    const currentTime = moment(schedule, "HH:mm");
    const prevTime = moment(allSchedule[index - 1], "HH:mm");
    return currentTime.isAfter(prevTime);
  });
};

export const validateSchedules = async (schedules) => {
  if (schedules.some((schedule) => schedule === "")) {
    return { isValid: false, warningType: "empty" };
  }

  if (areSchedulesInOrder(schedules)) {
    return { isValid: true, warningType: "" };
  } else {
    return { isValid: false, warningType: "passed" };
  }
};

export const getUTCTimeEpoch = (time, enable24HourTimers, hostData) => {
  const [hours, minutes] = enable24HourTimers
    ? time.split(":")
    : moment(time, "hh:mm A").format("HH:mm").split(":");
  const [day, month, year] = moment(hostData?.drugOrder?.scheduledDate)
    .format("DD-MM-YYYY")
    .split("-");
  const localTime = moment(
    `${year}-${month}-${day} ${hours}:${minutes}`,
    "YYYY-MM-DD HH:mm"
  );
  const utcTimeEpoch = moment.utc(localTime).unix();
  return utcTimeEpoch;
};
