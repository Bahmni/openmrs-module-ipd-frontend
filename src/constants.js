// Pick up output.publicPath from webpack.config.js. Maps to the proxy setting when using the micro-frontend capability
/* global __webpack_public_path__:writable */
export const BASE_URL =
  typeof __webpack_public_path__ !== "undefined"
    ? __webpack_public_path__
    : "/";
export const LS_LANG_KEY = "NG_TRANSLATE_LANG_KEY";

const hostUrl = localStorage.getItem("host")
  ? "https://" + localStorage.getItem("host")
  : "";
const RESTWS_V1 = hostUrl + "/openmrs/ws/rest/v1";
const FHIR2_R4 = hostUrl + "/openmrs/ws/fhir2/R4";
const BAHMNI_CORE = RESTWS_V1 + "/bahmnicore";

export const MEDICATIONS_BASE_URL = RESTWS_V1 + "/ipd/schedule/type/medication";
export const PATIENT_URL = RESTWS_V1 + "/patient";

export const DIAGNOSIS_SEARCH_URL = BAHMNI_CORE + "/diagnosis/search";
export const PRESCRIBED_AND_ACTIVE_DRUG_ORDERS_URL =
  BAHMNI_CORE + "/drugOrders/prescribedAndActive";

export const ALLERGIES_BASE_URL = FHIR2_R4 + "/AllergyIntolerance";
export const PATIENT_VITALS_URL = BAHMNI_CORE + "/diseaseSummaryData";
export const CLINICAL_CONFIG_URL =
  hostUrl + "/bahmni_config/openmrs/apps/clinical/app.json";

export const medicationFrequency = {
  START_TIME_DURATION_FREQUENCY: "START_TIME_DURATION_FREQUENCY",
  FIXED_SCHEDULE_FREQUENCY: "FIXED_SCHEDULE_FREQUENCY",
};
export const defaultDateFormat = "DD MMM YYYY";
export const defaultDateTimeFormat = "DD MMM YYYY hh:mm a";
export const DDMMYYY_DATE_FORMAT = "DD/MM/YYYY";
export const dateFormat = "DD/MM/YYYY";

export const componentKeys = {
  ALLERGIES: "AL",
  VITALS: "VT",
  DIAGNOSIS: "DG",
  TREATMENTS: "TR",
  NURSING_TASKS: "NT",
};
