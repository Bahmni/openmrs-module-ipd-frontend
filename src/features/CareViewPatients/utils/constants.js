export const searchKeys = ["bedNumber", "patientIdentifier", "patientName"];
export const items = [
    { id: "allTasks", text: "All Tasks" },
    { id: "medicationTasks", text: "Medication Tasks" },
    { id: "nonMedicationTasks", text: "Non-Medication Tasks" },
  ];
  