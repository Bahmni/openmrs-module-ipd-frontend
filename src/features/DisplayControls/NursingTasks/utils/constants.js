export const items = [
  { id: "allTasks", text: "All Tasks" },
  { id: "completed", text: "Completed" },
  { id: "pending", text: "Pending" },
  { id: "stopped", text: "Stopped" },
];
