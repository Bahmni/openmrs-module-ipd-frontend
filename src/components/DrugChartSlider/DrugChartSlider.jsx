import React, { useEffect, useState } from "react";
import {
  DatePickerCarbon,
  DropdownCarbon,
  NumberInputCarbon,
  TimePicker,
  TimePicker24Hour,
  Title,
} from "bahmni-carbon-ui";
import { TextArea, TextInput } from "carbon-components-react";
import moment from "moment";
import PropTypes from "prop-types";
import { FormattedMessage } from "react-intl";
import { I18nProvider } from "../../features/i18n/I18nProvider";
import { saveMedication } from "../../utils/DrugChartSliderUtils";
import SideBarPanel from "../../features/SideBarPanel/components/SideBarPanel";
import "./DrugChartSlider.scss";
import { SaveAndCloseButtons } from "../../features/SaveAndCloseButtons/components/SaveAndCloseButtons";
import {
  isInvalidTimeTextPresent,
  isValidSchedule,
  createDrugChartPayload,
  handleSchedule,
  handleStartTime,
} from "./DrugChartSliderUtils";

const DrugChartSlider = (props) => {
  const { title, hostData, hostApi, setDrugChartNotes, drugChartNotes } = props;
  const enableSchedule = hostData?.scheduleFrequencies.find(
    (frequency) =>
      frequency.name === hostData?.drugOrder?.uniformDosingType?.frequency
  );
  const enableStartTime = hostData?.startTimeFrequencies.includes(
    hostData?.drugOrder?.uniformDosingType?.frequency
  );
  const enable24HourTimers = hostData?.enable24HourTimers || false;
  const isAutoFill = !!enableSchedule?.scheduleTiming;

  const [schedules, setSchedules] = useState([]);
  const [startTime, setStartTime] = useState("");
  const [
    showStartTimeBeyondNextDoseWarning,
    setShowStartTimeBeyondNextDoseWarning,
  ] = useState(false);
  const [showStartTimePassedWarning, setShowStartTimePassedWarning] =
    useState(false);
  const [showSchedulePassedWarning, setShowSchedulePassedWarning] = useState(
    []
  );
  const [showScheduleOrderWarning, setShowScheduleOrderWarning] =
    useState(false);
  const [showEmptyScheduleWarning, setShowEmptyScheduleWarning] =
    useState(false);
  const [showEmptyStartTimeWarning, setShowEmptyStartTimeWarning] =
    useState(false);
  const invalidTimeText24Hour = "Please enter in 24-hr format";
  const invalidTimeText12Hour = "Please enter in 12-hr format";

  useEffect(() => {
    if (isAutoFill) {
      if (enable24HourTimers) {
        setSchedules(enableSchedule?.scheduleTiming || []);
      } else {
        const parsedTimings = enableSchedule?.scheduleTiming.map((time) =>
          moment(time, "hh:mm A")
        );
        setSchedules(parsedTimings || []);
      }
    } else {
      const defaultSchedules = Array.from(
        { length: enableSchedule?.frequencyPerDay },
        () => ""
      );
      setSchedules(defaultSchedules);
    }
  }, [isAutoFill, enable24HourTimers, enableSchedule]);

  const validateSave = async () => {
    if (
      isInvalidTimeTextPresent(
        enable24HourTimers,
        invalidTimeText12Hour,
        invalidTimeText24Hour
      )
    )
      return false;
    if (enableSchedule) {
      return await isValidSchedule(
        schedules,
        setShowEmptyScheduleWarning,
        setShowScheduleOrderWarning
      );
    }
    if (enableStartTime) {
      if (!startTime) {
        setShowEmptyStartTimeWarning(true);
        return false;
      }
      return true;
    }
  };

  const handleSave = async () => {
    const performSave = await validateSave();
    if (performSave) {
      const medication = createDrugChartPayload(
        hostData,
        drugChartNotes,
        enableStartTime,
        enableSchedule,
        enable24HourTimers,
        schedules,
        startTime
      );
      const response = await saveMedication(medication);
      response.status === 200 ? hostApi.onModalSave?.() : null;
    }
  };

  const handleCancel = () => {
    hostApi.onModalCancel?.();
  };

  const handleClose = () => {
    hostApi.onModalClose?.();
  };

  return (
    <I18nProvider>
      <SideBarPanel
        title={
          title ? title : <FormattedMessage id="DRUG_CHART_MODAL_HEADER" />
        }
        closeSideBar={handleClose}
      >
        <div style={{ padding: "20px", paddingBottom: "120px" }}>
          <TextInput
            id="drug-name"
            className="drug-name"
            type="text"
            value={hostData?.drugOrder?.drug?.name}
            labelText="Drug Name"
            disabled
          />
          <div className="inline-field">
            <div className="dose-field-with-units">
              <NumberInputCarbon
                id={"Dropdown"}
                onChange={() => {}}
                style={{ width: "50%" }}
                label={"Dose"}
                value={hostData?.drugOrder?.uniformDosingType?.dose}
                isDisabled={true}
              />
              <DropdownCarbon
                id={"Dropdown"}
                onChange={() => {}}
                titleText={" "}
                style={{ paddingLeft: "10px" }}
                selectedValue={
                  hostData?.drugOrder?.uniformDosingType?.doseUnits
                }
                options={[]}
                isDisabled={true}
              />
            </div>
            <div className="route">
              <DropdownCarbon
                id={"Dropdown"}
                onChange={() => {}}
                titleText={"Route"}
                selectedValue={hostData?.drugOrder?.route}
                options={[]}
                isDisabled={true}
              />
            </div>
          </div>
          <div className="inline-field">
            <div className="duration-field-with-units">
              <NumberInputCarbon
                id={"Dropdown"}
                onChange={() => {}}
                label={"Duration"}
                value={hostData?.drugOrder?.duration}
                isDisabled={true}
              />
              <DropdownCarbon
                id={"Dropdown"}
                onChange={() => {}}
                titleText={" "}
                selectedValue={hostData?.drugOrder?.durationUnit}
                options={[]}
                isDisabled={true}
              />
            </div>
            <DatePickerCarbon
              id={"Dropdown"}
              onChange={() => {}}
              titleText={"Start Date"}
              title={"Start Date"}
              value={moment(hostData?.drugOrder?.scheduledDate).format(
                "MM/DD/YYYY"
              )}
              isDisabled={true}
            />
          </div>
          <div className="frequency">
            <DropdownCarbon
              id={"DropdownFrequency"}
              onChange={() => {}}
              titleText={"Frequency"}
              selectedValue={hostData?.drugOrder?.uniformDosingType?.frequency}
              options={[]}
              isDisabled={true}
            />
          </div>
          {enableSchedule && (
            <div className="schedule-section">
              <Title text="Schedule(s)" isRequired={true} />
              <div className="inline-field" id="schedule">
                {Array.from(
                  { length: enableSchedule?.frequencyPerDay },
                  (_, index) =>
                    enable24HourTimers ? (
                      <div className="schedule-time" key={index}>
                        <TimePicker24Hour
                          key={index}
                          id={`schedule-${index}`}
                          defaultTime={schedules[index]}
                          onChange={(time) => {
                            handleSchedule(
                              time,
                              index,
                              schedules,
                              setSchedules,
                              enable24HourTimers,
                              invalidTimeText12Hour,
                              invalidTimeText24Hour,
                              setShowSchedulePassedWarning
                            );
                          }}
                          labelText=" "
                          width="70%"
                          invalidText={invalidTimeText24Hour}
                        />
                      </div>
                    ) : (
                      <div className="schedule-time" key={index}>
                        <TimePicker
                          key={index}
                          labelText=" "
                          defaultTime={schedules[index]}
                          onChange={(time) => {
                            handleSchedule(
                              time,
                              index,
                              schedules,
                              setSchedules,
                              enable24HourTimers,
                              invalidTimeText12Hour,
                              invalidTimeText24Hour,
                              setShowSchedulePassedWarning
                            );
                          }}
                          id={`schedule-${index}`}
                          invalidText={invalidTimeText12Hour}
                        />
                      </div>
                    )
                )}
              </div>
              {showScheduleOrderWarning && (
                <p className="time-error">
                  <FormattedMessage id="DRUG_CHART_MODAL_SCHEDULE_ORDER_WARNING"></FormattedMessage>
                </p>
              )}
              {showEmptyScheduleWarning && (
                <p className="time-error">
                  <FormattedMessage id="DRUG_CHART_MODAL_EMPTY_SCHEDULE_WARNING"></FormattedMessage>
                </p>
              )}
              {showSchedulePassedWarning.some(
                (showSchedulePassed) => showSchedulePassed === true
              ) && (
                <p className="time-warning">
                  <FormattedMessage id="DRUG_CHART_MODAL_SCHEDULE_PASSED_WARNING"></FormattedMessage>
                </p>
              )}
            </div>
          )}
          {enableStartTime && (
            <div className="start-time">
              {enable24HourTimers ? (
                <>
                  <TimePicker24Hour
                    data-modal-primary-focus
                    labelText={"Start Time"}
                    id={"start-time"}
                    onChange={(time) =>
                      handleStartTime(
                        time,
                        enable24HourTimers,
                        hostData,
                        setShowEmptyStartTimeWarning,
                        setShowStartTimeBeyondNextDoseWarning,
                        setShowStartTimePassedWarning,
                        setStartTime
                      )
                    }
                    isRequired={true}
                    defaultTime={startTime}
                    width="80%"
                    invalidText={invalidTimeText24Hour}
                  />
                  {showStartTimePassedWarning && (
                    <p className="time-warning">
                      <FormattedMessage id="DRUG_CHART_MODAL_START_TIME_PASSED"></FormattedMessage>
                    </p>
                  )}
                  {showStartTimeBeyondNextDoseWarning && (
                    <p className="time-warning">
                      <FormattedMessage id="DRUG_CHART_MODAL_START_TIME_BEYOND_NEXT_DOSE"></FormattedMessage>
                    </p>
                  )}
                  {showEmptyStartTimeWarning && (
                    <p className="time-error">
                      <FormattedMessage id="DRUG_CHART_MODAL_EMPTY_START_TIME_WARNING"></FormattedMessage>
                    </p>
                  )}
                </>
              ) : (
                <>
                  <TimePicker
                    id={"start-time"}
                    onChange={(time) =>
                      handleStartTime(
                        time,
                        enable24HourTimers,
                        hostData,
                        setShowEmptyStartTimeWarning,
                        setShowStartTimeBeyondNextDoseWarning,
                        setShowStartTimePassedWarning,
                        setStartTime
                      )
                    }
                    defaultTime={startTime}
                    labelText={"Start Time"}
                    isRequired={true}
                    invalidText={invalidTimeText12Hour}
                  />
                  {showStartTimePassedWarning && (
                    <p className="time-warning">
                      <FormattedMessage id="DRUG_CHART_MODAL_START_TIME_PASSED"></FormattedMessage>
                    </p>
                  )}
                  {showStartTimeBeyondNextDoseWarning && (
                    <p className="time-warning">
                      <FormattedMessage id="DRUG_CHART_MODAL_START_TIME_BEYOND_NEXT_DOSE"></FormattedMessage>
                    </p>
                  )}
                  {showEmptyStartTimeWarning && (
                    <p className="time-error">
                      <FormattedMessage id="DRUG_CHART_MODAL_EMPTY_START_TIME_WARNING"></FormattedMessage>
                    </p>
                  )}
                </>
              )}
            </div>
          )}
          <div className="instructions">
            <TextArea
              className="instruction"
              readOnly
              type="text"
              rows={1}
              value={hostData?.drugOrder?.instructions}
              labelText="Instruction"
              disabled
            />
          </div>
          <div className="additional-instructions">
            <TextArea
              className="additional-instruction"
              readOnly
              type="text"
              rows={1}
              value={hostData?.drugOrder?.additionalInstructions}
              labelText="Additional Instruction"
              disabled
            />
          </div>
          <div className="notes-sections">
            <TextArea
              data-testid="notes-section"
              className="notes-section"
              type="text"
              rows={3}
              value={drugChartNotes}
              onChange={(e) => setDrugChartNotes(e.target.value)}
              labelText="Notes"
            />
          </div>
        </div>
        <SaveAndCloseButtons
          onSave={() => handleSave()}
          onClose={() => handleCancel()}
        />
      </SideBarPanel>
    </I18nProvider>
  );
};

DrugChartSlider.propTypes = {
  title: PropTypes.string.isRequired,
  hostData: PropTypes.object.isRequired,
  hostApi: PropTypes.object.isRequired,
  setDrugChartNotes: PropTypes.func.isRequired,
  drugChartNotes: PropTypes.string.isRequired,
};

export default DrugChartSlider;
