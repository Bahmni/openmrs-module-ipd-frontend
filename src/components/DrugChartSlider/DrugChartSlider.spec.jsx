import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import DrugChartSlider from "./DrugChartSlider";
import {
  mockStartTimeDrugOrder,
  mockScheduleDrugOrder,
  mockScheduleFrequencies,
  mockStartTimeFrequencies,
  mockDrugOrderFrequencies,
  mockScheduleFrequenciesWithTimings,
} from "./DrugChartSliderTestUtils";
import "@testing-library/jest-dom";
import mockAdapter from "axios-mock-adapter";
import axios from "axios";
import { DRUG_ORDERS_CONFIG_URL } from "../../constants";
import DrugChartSliderNotification from "./DrugChartSliderNotification";

let mockAxios;

describe("DrugChartSlider", () => {
  beforeEach(() => {
    mockAxios = new mockAdapter(axios);
    const drugOrderFrequencies = mockDrugOrderFrequencies;
    mockAxios.onGet(DRUG_ORDERS_CONFIG_URL).reply(200, {
      results: drugOrderFrequencies,
    });
  });

  afterEach(() => {
    mockAxios.reset();
  });

  it("Component renders successfully", async () => {
    render(
      <DrugChartSlider
        hostData={{
          drugOrder: mockStartTimeDrugOrder,
          scheduleFrequencies: mockScheduleFrequencies,
          startTimeFrequencies: mockStartTimeFrequencies,
        }}
        hostApi={{}}
      />
    );
    await waitFor(() => {
      expect(screen.getByText("Add to Drug Chart")).toBeTruthy();
    });
  });

  it("should show drug name field to be disabled", async () => {
    render(
      <DrugChartSlider
        hostData={{
          drugOrder: mockScheduleDrugOrder,
          scheduleFrequencies: mockScheduleFrequencies,
          startTimeFrequencies: mockStartTimeFrequencies,
        }}
        hostApi={{}}
      />
    );
    await waitFor(() => {
      const inputElement = screen.getByLabelText("Drug Name");
      expect(inputElement).toBeInTheDocument();
      expect(inputElement).not.toHaveStyle("cursor: pointer");
    });
  });

  it("should show dose field to be disabled", async () => {
    render(
      <DrugChartSlider
        hostData={{
          drugOrder: mockScheduleDrugOrder,
          scheduleFrequencies: mockScheduleFrequencies,
          startTimeFrequencies: mockStartTimeFrequencies,
        }}
        hostApi={{}}
      />
    );
    await waitFor(() => {
      const inputElement = screen.getByLabelText("Dose");
      expect(inputElement).toBeInTheDocument();
      expect(inputElement).not.toHaveStyle("cursor: pointer");
    });
  });

  it("should show duration field to be disabled", async () => {
    render(
      <DrugChartSlider
        hostData={{
          drugOrder: mockScheduleDrugOrder,
          scheduleFrequencies: mockScheduleFrequencies,
          startTimeFrequencies: mockStartTimeFrequencies,
        }}
        hostApi={{}}
      />
    );
    await waitFor(() => {
      const inputElement = screen.getByLabelText("Duration");
      expect(inputElement).toBeInTheDocument();
      expect(inputElement).not.toHaveStyle("cursor: pointer");
    });
  });

  it("should show start date field to be disabled", async () => {
    render(
      <DrugChartSlider
        hostData={{
          drugOrder: mockScheduleDrugOrder,
          scheduleFrequencies: mockScheduleFrequencies,
          startTimeFrequencies: mockStartTimeFrequencies,
        }}
        hostApi={{}}
      />
    );
    await waitFor(() => {
      const inputElement = screen.getByLabelText("Start Date");
      expect(inputElement).toBeInTheDocument();
      expect(inputElement).not.toHaveStyle("cursor: pointer");
    });
  });

  it("should show notes field to be enabled", async () => {
    render(
      <DrugChartSlider
        hostData={{
          drugOrder: mockStartTimeDrugOrder,
          scheduleFrequencies: mockScheduleFrequencies,
          startTimeFrequencies: mockStartTimeFrequencies,
        }}
        hostApi={{}}
      />
    );
    await waitFor(() => {
      const inputElement = screen.getByTestId("notes-section");
      expect(inputElement).toBeInTheDocument();
      expect(inputElement).not.toBeDisabled();
    });
  });

  it("should enable schedule when frequency is present in scheduleFrequencies", async () => {
    render(
      <DrugChartSlider
        hostData={{
          scheduleFrequencies: mockScheduleFrequencies,
          startTimeFrequencies: mockStartTimeFrequencies,
          drugOrder: mockScheduleDrugOrder,
        }}
        hostApi={{}}
      />
    );
    await waitFor(() => {
      const inputElement = screen.getByText("Schedule(s)");
      expect(inputElement).toBeTruthy();
    });
  });

  it("should enable start time when frequency is present in scheduleFrequencies", async () => {
    render(
      <DrugChartSlider
        hostData={{
          scheduleFrequencies: mockScheduleFrequencies,
          startTimeFrequencies: mockStartTimeFrequencies,
          drugOrder: mockStartTimeDrugOrder,
        }}
        hostApi={{}}
      />
    );
    await waitFor(() => {
      const inputElement = screen.getByText("Start Time");
      expect(inputElement).toBeTruthy();
    });
  });

  it("should show Please select Schedule(s) when save is clicked without entering schedule", async () => {
    render(
      <DrugChartSlider
        hostData={{
          scheduleFrequencies: mockScheduleFrequencies,
          startTimeFrequencies: mockStartTimeFrequencies,
          drugOrder: mockScheduleDrugOrder,
        }}
        hostApi={{}}
      />
    );
    await waitFor(() => {
      const saveButton = screen.getByRole("button", { name: "Save" });
      fireEvent.click(saveButton);
    });
    expect(screen.getByText("Please enter Schedule(s)")).toBeInTheDocument();
  });

  it("should show Please select Start Time when save is clicked without entering start time", async () => {
    render(
      <DrugChartSlider
        hostData={{
          scheduleFrequencies: mockScheduleFrequencies,
          startTimeFrequencies: mockStartTimeFrequencies,
          drugOrder: mockStartTimeDrugOrder,
        }}
        hostApi={{}}
      />
    );
    await waitFor(() => {
      const saveButton = screen.getByRole("button", { name: "Save" });
      fireEvent.click(saveButton);
    });
    expect(screen.getByText("Please enter Start Time")).toBeInTheDocument();
  });

  it("Should show invalid time format error message when wrong time is entered in the field", async () => {
    render(
      <DrugChartSlider
        hostData={{
          enable24HourTimers: true,
          scheduleFrequencies: mockScheduleFrequencies,
          startTimeFrequencies: mockStartTimeFrequencies,
          drugOrder: mockStartTimeDrugOrder,
        }}
        hostApi={{}}
      />
    );

    await waitFor(() => {
      const startTimeInput = document.getElementById("time-selector");
      const timeValue = "41:22";
      fireEvent.change(startTimeInput, { target: { value: timeValue } });
    });

    fireEvent.blur(document.getElementById("time-selector"));
    expect(
      screen.getByText("Please enter in 24-hr format")
    ).toBeInTheDocument();
  });

  it("Should show pre-filled timing in the schedule fields if the schedule time is provided from config", async () => {
    render(
      <DrugChartSlider
        hostData={{
          enable24HourTimers: true,
          scheduleFrequencies: mockScheduleFrequenciesWithTimings,
          startTimeFrequencies: mockStartTimeFrequencies,
          drugOrder: mockScheduleDrugOrder,
        }}
        hostApi={{}}
      />
    );

    await waitFor(() => {
      const startTimeInputs = document.querySelectorAll("#time-selector");
      expect(startTimeInputs).toBeDefined();
      expect(startTimeInputs[0].value).toBe("8:00");
      expect(startTimeInputs[1].value).toBe("16:00");
    });
  });
});

describe("DrugChartSliderNotification", () => {
  it("Component renders successfully", async () => {
    const { container } = render(
      <DrugChartSliderNotification hostData={{}} hostApi={{}} />
    );
    expect(container).toMatchSnapshot();
  });
});
