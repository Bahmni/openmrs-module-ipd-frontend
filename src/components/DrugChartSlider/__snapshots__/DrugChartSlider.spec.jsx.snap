// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrugChartSliderNotification Component renders successfully 1`] = `
<div>
  <div
    class="alertContainer"
  >
    <div>
      <div
        class="bx--inline-notification bx--inline-notification--low-contrast bx--inline-notification--success bx--inline-notification--hide-close-button"
        kind="success"
        role="alert"
      >
        <div
          class="bx--inline-notification__details"
        >
          <svg
            aria-hidden="true"
            class="bx--inline-notification__icon"
            fill="currentColor"
            focusable="false"
            height="20"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10,1c-4.9,0-9,4.1-9,9s4.1,9,9,9s9-4,9-9S15,1,10,1z M8.7,13.5l-3.2-3.2l1-1l2.2,2.2l4.8-4.8l1,1L8.7,13.5z"
            />
            <path
              d="M8.7,13.5l-3.2-3.2l1-1l2.2,2.2l4.8-4.8l1,1L8.7,13.5z"
              data-icon-path="inner-path"
              fill="none"
              opacity="0"
            />
            <title>
              success icon
            </title>
          </svg>
          <div
            class="bx--inline-notification__text-wrapper"
          >
            <p
              class="bx--inline-notification__title"
            >
              title
            </p>
            <div
              class="bx--inline-notification__subtitle"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
